pallet-ismp/src/evm/handler.rs

```diff
@@ -0,0 +1,202 @@
+//! Module Handler for EVM contracts
+use crate::{
+    evm::abi::{
+        ContractData as SolContractData, GetRequest as SolGetRequest,
+        GetResponse as SolGetResponse, OnAcceptCall, OnGetResponseCall, OnGetTimeoutCall,
+        OnPostResponseCall, OnPostTimeoutCall, OptionValue, PostRequest,
+        PostResponse as SolPostResponse, StorageValue as SolStorageValue,
+    },
+    primitives::ModuleId,
+    Config, GasLimits,
+};
+use alloy_primitives::U256;
+use alloy_sol_types::{SolCall, SolType};
+use core::marker::PhantomData;
+use ismp_rs::{
+    contracts::Gas,
+    error::Error,
+    module::IsmpModule,
+    router::{Post, Request, Response},
+};
+use pallet_evm::GasWeightMapping;
+use sp_core::H160;
+
+/// Handler host address
+/// Contracts should only allow ismp module callbacks to be executed by this address
+pub const EVM_HOST_ADDRESS: H160 = H160::zero();
+/// EVM contract handler
+pub struct EvmContractHandler<T: Config + pallet_evm::Config>(PhantomData<T>);
+
+impl<T: Config + pallet_evm::Config> Default for EvmContractHandler<T> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T: Config + pallet_evm::Config> IsmpModule for EvmContractHandler<T> {
+    fn on_accept(&self, request: Post) -> Result<Gas, Error> {
+        let target_contract = parse_contract_id(&request.to)?;
+        let contract_data = SolContractData::decode(&request.data, true).map_err(|_| {
+            Error::ImplementationSpecific(
+                "Failed to decode request data to the standard format".to_string(),
+            )
+        })?;
+        let gas_limit = contract_data.gasLimit;
+        let post = PostRequest {
+            source: request.source.to_string().as_bytes().to_vec(),
+            dest: request.dest.to_string().as_bytes().to_vec(),
+            nonce: u64_to_u256(request.nonce)?,
+            timeoutTimestamp: u64_to_u256(request.timeout_timestamp)?,
+            from: request.from,
+            to: request.to,
+            data: request.data,
+        };
+        let call_data = OnAcceptCall { request: post }.encode();
+        execute_call::<T>(target_contract, call_data, gas_limit)
+    }
+
+    fn on_response(&self, response: Response) -> Result<Gas, Error> {
+        let target_contract = parse_contract_id(&response.destination_module())?;
+        let gas_limit = GasLimits::<T>::get(response.nonce())
+            .ok_or(Error::ImplementationSpecific("Gas limit not found".to_string()))?;
+        GasLimits::<T>::remove(response.nonce());
```

Comment:
thought this was only for get requests?

Reply:
hmm, let's do it this way, when dispatching requests, apps should set `gas_limit` to `max(destination_gas_limit, response_gas_limit)`